protocol/consistencychecks.go

```diff
@@ -222,43 +241,53 @@ func (cc *ConsistencyChecks) verifyReturnedPromise(tb *m.TemporaryBinding,
 			return ErrorBadIndex
 		}
 
-		cc.TBs = append(cc.TBs, tb)
+		// verify TB's value
+		if !bytes.Equal(tb.Value, key) {
+			return ErrorBadPromise
+		}
+
+		cc.TBs[uname] = tb
 	}
 
 	return nil
 }
 
 // verifyFulfilledPromise verifies issued TBs was inserted
 // in the directory as promised.
-func (cc *ConsistencyChecks) verifyFulfilledPromise(str *m.SignedTreeRoot,
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
 	ap *m.AuthenticationPath) error {
-	if cc.isUseTBs {
-		backed := cc.TBs[:0]
-		for _, tb := range cc.TBs {
-			if cc.CurrentEpoch+1 == str.Epoch {
+	if cc.useTBs {
+		for name, tb := range cc.TBs {
+			if cc.CurrentEpoch+1 == str.Epoch &&
+				name == uname {
 				if !tb.Verify(ap) {
 					return ErrorBrokenPromise
 				}
-			} else if cc.CurrentEpoch == str.Epoch {
-				// keep current epoch's returned promises
-				// for future verifications
-				backed = append(backed, tb)
+				delete(cc.TBs, uname)
+			} else if cc.CurrentEpoch < str.Epoch {
+				// clear all issued promises since they have been verified
+				// or the client has missed some epochs
```

Comment:
> For our first release, deleting old TBs seems fine, since we're only supporting operations in the same epoch, but when we introduce KeyLookupInEpoch and monitoring with missed epochs, we should reconsider this.

Is that handled in our previous discussion (https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84501211)?